examples/control-vector-generator/control-vector-generator.cpp

```diff
@@ -0,0 +1,763 @@
+    printf("Done with PCA.\n");
+}
+
+template <typename T>
+static std::string to_string(const T & val) {
+    std::stringstream ss;
+    ss << val;
+    return ss.str();
+}
+
+static void export_gguf(callback_data & cb_data, int n_layers, const std::string fname, const std::string model_hint) {
+    struct gguf_context * ctx = gguf_init_empty();
+
+    size_t v_final_size_eff = n_layers - 1;
+    
+    const std::string arch = "controlvector";
+    gguf_set_val_str(ctx, "general.architecture", arch.c_str());
+    gguf_set_val_str(ctx, (arch + ".model_hint").c_str(), model_hint.c_str());
+    gguf_set_val_i32(ctx, (arch + ".layer_count").c_str(), v_final_size_eff);
+
+    struct ggml_init_params params = {
+        /*.mem_size   =*/ (ggml_tensor_overhead() * v_final_size_eff)
+                            + (cb_data.n_embd * v_final_size_eff * sizeof(float)),
+        /*.mem_buffer =*/ NULL,
+        /*.no_alloc   =*/ false,
+    };
+
+    struct ggml_context * ctx_data = ggml_init(params);
+
+    for (size_t i = 0; i < v_final_size_eff; ++i) {
+        // TODO this number is probably not right - figure out which layer is which
+        // the python implementation uses a dict to handle this, we don't know if it's 1, 2, 3, 4... or other
+        const std::string name = "direction." + to_string(i+1);
+
+        struct ggml_tensor * cur = ggml_new_tensor_1d(ctx_data, GGML_TYPE_F32, cb_data.n_embd);
+
+        ggml_set_name(cur, name.c_str());
+
+        float * data = (float *) cur->data;
+        for(int j = 0; j < cb_data.n_embd; j++) {
+            data[j] = cb_data.v_final[i][j];
+        }
+
+        gguf_add_tensor(ctx, cur);
+        printf("Added tensor %zu\n", i);
+    }
+
+    printf("Writing file...\n"); 
+
+    gguf_write_to_file(ctx, fname.c_str(), false);
+
+    printf("%s: wrote file '%s'\n", __func__, fname.c_str());
+
+    ggml_free(ctx_data);
+    gguf_free(ctx);
+}
+
+// END NON-GGML IMPLEMENTATION
+
+int main(int argc, char ** argv) {
+    ctrl_params cparams;
+
+    int skipme = ctrlvec_params_parse(argc, argv, cparams);
+    argc -= skipme;
+    argv += skipme;
+
+    gpt_params params;
+    if (!gpt_params_parse(argc, argv, params)) {
+        return 1;
+    }
+
+    // load prompts
+    cparams.positive_prompts = ctrlvec_load_prompt_file(cparams.positive_prompts_file);
+    cparams.negative_prompts = ctrlvec_load_prompt_file(cparams.negative_prompts_file);
+    if (cparams.positive_prompts.size() != cparams.negative_prompts.size()) {
+        fprintf(stderr, "number of positive and negative prompts must be equal");
+        return 1;
+    }
+    if (cparams.positive_prompts.empty()) {
+        fprintf(stderr, "must provide at least one prompt pair");
+        return 1;
+    }
+
+    callback_data cb_data;
+
+    // pass the callback to the backend scheduler
+    // it will be executed for each node during the graph computation
+    params.cb_eval = cb_eval;
+    params.cb_eval_user_data = &cb_data;
```

Comment:
We can't set `params.cb_eval_user_data` to another value once the model is loaded. That's why in my previous version, I reload the model every time a new `cb_data` is created.

However, here we can do better: instead of creating a new `cb_data`, we can maybe clear the old one and reuse it.

Reply:
The `cb_eval` is part of the context params, so you would only need to create a new context, but you can reuse the same model. Adding a function to the llama API to change it would also be possible. But yeah, reusing the same one would probably be more efficient.

Reply:
I changed your previous version so that we only create one `cb_data` object for the entire run of the program, but this might also be undesirable since I abuse it a little (put more in there than there should be). Later we can look into refactoring the `callback_data` struct when we finalize the GGML implementation.